rust/kernel/error.rs

```diff
@@ -58,6 +58,26 @@ impl Error {
     pub fn to_kernel_errno(&self) -> c_types::c_int {
         self.0
     }
+
+    /// Creates an [`Error`] from a kernel pointer.
+    ///
+    /// # Pointer value range
+    ///
+    /// According to `include/linux/err.h`:
+    /// - [0, .., `core::usize::MAX - bindings::MAX_ERRNO`) is the range for
+    ///   normal values of pointer.
+    ///
+    /// - [`core::usize::MAX - bindings::MAX_ERRNO`,..,`core::usize::MAX] is the
+    ///   range for error values stored in pointer.
+    pub fn from_kernel_ptr(ptr: *mut c_types::c_void) -> Option<Error> {
+        let value = ptr as usize;
+
+        if value >= core::usize::MAX - bindings::MAX_ERRNO as usize {
+            Some(Error::from_kernel_errno(value as c_types::c_int))
```

Comment:
Is casting pointers to `c_int` idiomatic? And aren't some pointers wider than `c_int`? Could we perhaps use the C function provided by the kernel, i.e. `PTR_ERR()`? We know it'll always work.

Reply:
We have done the check on `value`, so it's guaranteed that it's safe to convert to `c_int`. As for using `PTR_ERR()`, as I raised in another PR, we would have extra cost on function calls.